### ISSUE TITLE
Approaches for using fixtures with unit tests

### ISSUE THREAD

**Issue by mjrusso** [bitcrowd/carbonite#79]
I'm currently evaluating Carbonite (the library is extremely elegant, by the way -- thank you!).

Ideally I'd like my unit tests to catch any cases where the `Carbonite.Transaction` is missing, but am running into some difficulty here. In more detail:

I set up Carbonite and added a trigger to a single table (in a Phoenix 1.7 project, in the off chance that matters), and then immediately ran the unit tests. The test that creates a new record in that table fails:

``` bash
** (Postgrex.Error) ERROR 23503 (foreign_key_violation) INSERT on table public.foos without prior INSERT into carbonite_default.transactions
```

Perfect, exactly what I was hoping to see. I updated the code that inserts a new record to call `Carbonite.Multi.insert_transaction` as part of the transaction, and re-ran the tests.

That test now passes, but I also have some other failures related to tests that update and delete from this table. The underlying error is the same as above, but the difference is that these tests use a fixture that first creates a record (so that the tests can modify or delete the record and then assert on the outcome).

I updated the test fixture to call `Carbonite.Multi.insert_transaction` as part of the transaction, and re-ran the tests.

However, they all passed... which is not what I wanted to see, because my update and delete context methods haven't been updated yet. The unit tests pass, but the production app crashes when updating or deleting records from this table.

I tried modifying the fixture to call `Carbonite.Multi.override_mode` instead of `Carbonite.Multi.insert_transaction`, like so:

``` elixir
Ecto.Multi.new()
|> Carbonite.Multi.override_mode()
|> Ecto.Multi.insert(:foo, foo)
|> Repo.transaction()
```

However, the tests still pass with this change.

From [the docs](https://hexdocs.pm/carbonite/Carbonite.html#insert_transaction/3-multiple-inserts-in-the-same-transaction):

> In tests using Ecto's SQL sandbox, subsequent calls to transactional operations (even to the same operation twice) are wrapped inside the overarching test transaction, and hence also effectively call `insert_transaction/3` within the same transaction.
>
> ...
>
> Therefore, `insert_transaction/3` ignores subsequent calls within the same database transaction (equivalent to `ON CONFLICT DO NOTHING`), discarding metadata passed to all calls but the first.

I'm assuming that the behaviour I'm seeing is because of the above?

Digging around in [the source](https://github.com/bitcrowd/carbonite/blob/3e9e2c3f054a54ab1eaf5cb3ac00245ff57819e4/lib/carbonite/query.ex#L50-L73), I came across:

``` elixir
@doc """
Returns an `t:Ecto.Query.t/0` that can be used to select or delete the "current" transaction.

This function is useful when your tests run in a database transaction using Ecto's SQL sandbox.

## Example: Asserting on the current transaction

When you insert your `Carbonite.Transaction` record somewhere inside your domain logic, you do
not wish to return it to the caller only to be able to assert on its attributes in tests. This
example shows how you could assert on the metadata inserted.

    # Test running inside Ecto's SQL sandbox.
    test "my test" do
      some_operation_with_a_transaction()
      assert current_transaction_meta() == %{"type" => "some_operation"}
    end

    defp current_transaction_meta do
      Carbonite.Query.current_transaction()
      |> MyApp.Repo.one!()
      |> Map.fetch(:meta)
    end

...

"""
```

If I use `override_mode` when creating a record (in the fixture), then `Carbonite.Query.current_transaction() |> Repo.one!()` fails with:

```
** (Ecto.NoResultsError) expected at least one result but got none in query:

     from t0 in Carbonite.Transaction,
       prefix: "carbonite_default",
       where: t0.id == fragment("CURRVAL(CONCAT(?::VARCHAR, '.transactions_id_seq'))", ^"carbonite_default"),
       where: t0.xact_id == fragment("pg_current_xact_id()")
```

So I could assert on `Carbonite.Query.current_transaction()` returning a result in my tests, but I'd rather (with the exception of the fixtures) not need for them to know anything about Carbonite.

Which brings me to my questions... am I doing something wrong? Is there a better way to approach this?

**Comment by andreasknoepfle** [comment]
Hello @mjrusso,

yes, you are right in transactional tests with ecto's sandbox, when you call `insert_transaction` multiple times nothing will happen. 

When using   `override_mode`, keep in mind, that after calling it once for the fixtures in a test transaction it will not only be active for your fixtures, but for the rest of the test case.

> If you are using Ecto's SQL sandbox for running transactional tests, this means the override is going to be active until the end of the test case. -- https://hexdocs.pm/carbonite/Carbonite.html#module-testing-bypassing-carbonite

The reason for that [you can see in the trigger itself](
https://github.com/bitcrowd/carbonite/blob/160f0c00fceaf6144e4fae21593661de566be4da/lib/carbonite/migrations/v5.ex#L34-L39). The trigger will look at the `mode` and will check if the current transaction will override this mode (essentially by setting `override_xact_id` on the trigger, which will be rolled back only at the end of the test transaction). 

This also means that when you set your `mode` to `capture` on the trigger, after calling `override_mode` once it will be `ignore` for the **rest of the transaction**. If you set your mode to `ignore`in the trigger, it will be `capture` after calling override mode for the **rest of the transaction**. 

So our general advice for using carbonite with ecto sandbox is starting out with `ignore` and then calling `override_mode`, once you are done with the setup (so the factories/fixtures will be ignored). 

In our codebase we even wrote a little helper for asserting a carbonite transaction was called in a certain way:

```
 @spec assert_carbonite_transaction(map(), function()) :: Carbonite.Transaction.t()
  def assert_carbonite_transaction(meta \\ %{}, fun) do
    Carbonite.override_mode(MyRepo)

    fun.()

    Carbonite.Query.current_transaction()
    |> MyRepo.one!()
    |> tap(fn transaction ->
      assert transaction.meta == meta
    end)
  end
```
We haven't generalised it yet enough to make it to carbonite, but we could eventually.

So you would do something like:
```
# config/test.exs
config :my_app, carbonite_mode: :ignore
```

And in the test:
```
setup :insert_fixtures

test "it inserts carbonite transactions" do
  assert_carbonite_transaction fn ->
     my_context_function()
  end
end
```

I hope that helps
👋

Cheers Andi

**Comment by maltoe** [comment]
Our usual setup as @andreasknoepfle described trades the risk of the occassional slip through of a missed `insert_transaction` for the freedom to ignore carbonite in _most_ tests. Although that slip (IIRC) has never happened to us in production, as usually while working on a feature you test your code in `:dev` and hence notice when you have missed the `insert_transaction`, which in turn (usually) reminds you to write a test with `assert_carbonite_transaction`.

> Ideally I'd like my unit tests to catch any cases where the Carbonite.Transaction is missing

Our approach is equivalent to option 2 in this [section in the docs](https://hexdocs.pm/carbonite/Carbonite.html#module-testing-bypassing-carbonite). If you want to go for option 1 instead to be extra sure you never miss an `insert_transaction`, you would need to bypass Carbonite when inserting your fixtures. Currently, we don't have a way to "reset" the override mode, but we can add one easily, so you could write something like this:

```elixir
defp my_setup_block(_) do
  with_carbonite_override_mode(fn -> insert_my_fixture() end)
end

defp with_carbonite_override_mode(fun) do
  Carbonite.override_mode()
  x = fun.()
  Carbonite.reset_override_mode()
  x
end
```

**Comment by mjrusso** [comment]
@andreasknoepfle and @maltoe, thank you both so much for taking the time to explain and for #80. I had read the "Testing / Bypassing Carbonite" section in the docs a few times but didn't completely grok it; I now understand this much more clearly.

I've updated my code to use #80, in conjunction with the "Bypassing change capture selectively" option and the `with_ignored_carbonite` helper, and it works brilliantly. I'll likely also adopt the `assert_carbonite_transaction` helper that @andreasknoepfle mentioned above in the cases where I want to be extra sure that the correct metadata was written.

Thanks again!

**Comment by maltoe** [comment]
released as v0.8.0
